core/src/queue.rs

```diff
@@ -220,64 +219,53 @@ impl Queue {
-                Entry::Occupied(entry) => entry,
-                // As practice shows this code is not `unreachable!()`.
-                // When transactions are submitted quickly it can be reached.
-                Entry::Vacant(_) => continue,
-            };
-            if self.check_tx(entry.get(), wsv).is_err() {
-                entry.remove_entry();
-                continue;
-            }
-
-            if *entry
-                .get()
-                .check_signature_condition(wsv)
-                .expect("Checked in `check_tx` just above")
-            {
-                return Some(entry.get().clone());
-            }
-        }
-    }
-
-    /// Return the number of transactions in the queue
+    /// Return the number of transactions in the queue.
     pub fn tx_len(&self) -> usize {
         self.txs.len()
     }
 
     /// Gets transactions till they fill whole block or till the end of queue.
     ///
     /// BEWARE: Shouldn't be called in parallel with itself.
-    #[allow(clippy::missing_panics_doc, clippy::unwrap_in_result)]
     pub fn get_transactions_for_block(
         &self,
         wsv: &WorldStateView,
     ) -> Vec<VersionedAcceptedTransaction> {
+        let mut transactions = Vec::with_capacity(self.txs_in_block);
+        self.put_transactions_for_block_into(wsv, &mut transactions);
+        transactions
+    }
+
+    /// Put transactions into provided vector until they fill the whole block or there are no more transactions in the queue.
+    ///
+    /// BEWARE: Shouldn't be called in parallel with itself.
+    pub fn put_transactions_for_block_into(
+        &self,
+        wsv: &WorldStateView,
+        transactions: &mut Vec<VersionedAcceptedTransaction>,
+    ) {
+        if transactions.len() >= self.txs_in_block {
+            return;
+        }
+
         let mut seen = Vec::new();
 
+        let mut transactions_hashes: HashSet<HashOf<VersionedSignedTransaction>> = transactions
+            .iter()
+            .map(VersionedAcceptedTransaction::hash)
+            .collect();
         let out = std::iter::repeat_with(|| self.pop(&mut seen, wsv))
-            .take_while(Option::is_some)
-            .map(Option::unwrap)
-            .take(self.txs_in_block)
-            .collect::<Vec<_>>();
+            .map_while(core::convert::identity)
+            // Normally we would not expect to receive transaction duplicate from the queue and `transaction_hashes.contains(tx.hash())` would be enough
+            // However there might be situation when we poped transaction from the queue and just after that we received its duplicate
+            .filter(|tx| transactions_hashes.insert(tx.hash()))
+            .take(self.txs_in_block - transactions.len());
+        transactions.extend(out);
 
         #[allow(clippy::expect_used)]
         seen.into_iter()
             .try_for_each(|hash| self.queue.push(hash))
             .expect("As we never exceed the number of transactions pending");
```

Comment:
Unrelated to this PR, but I really don't get where this guarantee comes from.

Reply:
I believe such guarantee came from the fact that we never pushing back more transactions that was poped previously (so there is should be enough space for them).

Reply:
Another thread could fill up the queue in the meantime though

Reply:
I suppose not because we don't remove transaction from `txs` map and inside `push` we have check for `txs` size.

Reply:
Oh, so we don't remove signed txs from the map until they're in blockchain. Got it, thanks!